### ISSUE TITLE
We need a better upgrade strategy - every update seems painful

### ISSUE THREAD

**Issue by ss89** [Mailu/Mailu#2316]
Thank you for opening an issue with Mailu. Please understand that issues are meant for bugs and enhancement-requests.
For **user-support questions**, reach out to us  on [matrix](https://matrix.to/#/#mailu:tedomum.net).

To be able to help you best, we need some more information.

## Before you open your issue
- [x] Check if no issue or pull-request for this already exists.
- [x] Check [documentation](https://mailu.io/master/) and [FAQ](https://mailu.io/master/faq.html). (Tip, use the search function on the documentation page)
- [x] You understand `Mailu` is made by volunteers in their **free time** — be conscise, civil and accept that delays can occur.
- [x] The title of the issue should be short and simple. It should contain specific terms related to the actual issue. Be specific while writing the title.

### Environment
 - [x] docker-compose
 - [ ] kubernetes
 - [ ] docker swarm

### Versions
1.9

## Description
Whenever i update my mailu setup it seems to break. 

This is the case since version 1.5.

I am very careful when reading the docs (which are quite limited) and when adjusting the docker-compose.yml/env files, to not introduce any issues myself.

However whenever i have/want to update my mailu setup, stuff seems to break. I believe this project needs to be better downward compatible.

For example: i tried to upgrade my mailu from 1.8 to 1.9 just now and what i got was http 500 on the login page of roundcube and 502 on the admin page. I checked the logs of roundcube and it didn't even seem to get the request, though the front nginx config clearly referred to the correct ip. I thought "ok, shit happened again, let's roll back", but then i was greeted with a login page that told me that the storage could not be reached - wait WHAT? Ofcourse the service was up and running and also the configuration was all good. It was just that the new database was not compatible anymore with the old software version. Luckily i had a backup at hand which i could quickly use to overrride the current db. Afterwards everything works again - on 1.8.

I do see that this is an OSS project and that everybody tries their best, but i would give a BIG upvote for a better upgrade experience. Every upgrade seems to add fear, because i am not into this whole project to quickly figure out what is wrong and at the same time i want my mail server to be offline as short as possible.

## Replication Steps
Try to upgrade from 1.8 to 1.9

## Expected behaviour
Working roundcube.

## Logs
````markdown
WARNING: Changed defaults (These config options have new default values):
- 'skin'
- 'smtp_port'
- 'smtp_user'
- 'smtp_pass'
- 'jquery_ui_skin_map'
Executing database schema update.
Updating database schema (2020020100)... [OK]
Updating database schema (2020020101)... [OK]
Updating database schema (2020091000)... [OK]
Updating database schema (2020122900)... [OK]
Indexing contacts for user 1...
Indexing contacts for user 2...
Indexing contacts for user 3...
Indexing contacts for user 4...
done.
This instance of Roundcube is up-to-date.
Have fun!
0 records deleted from 'contactgroupmembers'
0 records deleted from 'contacts'
0 records deleted from 'contactgroupmembers'
0 records deleted from 'contactgroups'
0 records deleted from 'identities'
[Sun Apr 10 16:10:46.963637 2022] [mpm_prefork:notice] [pid 14] AH00163: Apache/2.4.53 (Debian) configured -- resuming normal operations
[Sun Apr 10 16:10:46.963685 2022] [core:notice] [pid 14] AH00094: Command line: 'apache2 -D FOREGROUND'
````

**TLDR**
* upgrades should be easier
* it should be well explained what changed between software versions, referring to a bunch of files is not exactly nice
* database schema upgrades shouldn't stop you from downgrading

**Comment by baskinsy** [comment]
I had issues even with the simple miner version update. Executing ```docker-compose pull && docker-compose up -d``` has left my install in a non-working state, returning 502 error on web interface. Although it is fixed with a ```docker-compose stop``` followed by ```docker-compose start```, it has happened twice at least to me. It shouldn't break even like that with a minor update.

**Comment by nextgens** [comment]
> I had issues even with simple minor version updates. Executing `docker-compose pull && docker-compose up -d` has left my install in a non-working state, returning 502 error on web interface. Although it is fixed with a `docker-compose stop` followed by `docker-compose start`, it has happened twice at least to me. It shouldn't break even like that with a minor update.

The upgrade instructions at https://mailu.io/1.9/maintain.html?highlight=update#upgrading-the-mail-server need changing; It's well known that unless you stop the containers it won't work

**Comment by nextgens** [comment]
The fact that rollbacks to 1.8 is impossible is not due to a database schema change... it's due to the changes related to how user passwords are stored. If you upgrade and run with the new code the hashes get replaced with something the old code can't handle.

I don't think that it's reasonable for us to try to guarantee that newer versions of the database should work with older code and it's definitely not something that was attempted before we've adopted semantic versioning. If you meant to restore a backup you should have gone all the way and restore both code & database to what was working (and that would have worked!).

I take your point that the upgrade process should be better tested/documented. It would be great if we had a regression test for it; We'd welcome any patch.

**Comment by nextgens** [comment]
FYI We've discussed it on #mailu-dev and I will bring it up at our next scheduled meeting.

**Comment by baskinsy** [comment]
> The upgrade instructions at https://mailu.io/1.9/maintain.html?highlight=update#upgrading-the-mail-server need changing; It's well known that unless you stop the containers it won't work

The notification email also for new releases doesn't mention anything. 

>This is a new automatic release of Mailu. The new version can be seen in the tag name.
>The main version X.Y (e.g. 1.9) will always reflect the latest version of the branch. To update your Mailu installation simply pull the latest images docker-compose pull && docker-compose up -d.


Should be corrected also.

**Comment by nextgens** [comment]
Good point; I'll ensure I get it updated in the next PR we merge

**Comment by CryptosWORLD** [comment]
where can we find the upgrade steps for the latest version ?
thank you
